lib/msprime.c

```diff
@@ -2905,74 +2910,156 @@ msp_change_label(msp_t *self, segment_t *ind, label_id_t label)
     uint32_t sweep_locus = model->params.single_sweep.locus;
-    size_t j;
-    /* double *time = model->params.single_sweep.trajectory.time; */
-
+    double sweep_dt;
+    size_t j = 0;
+    int64_t num_links;
+    unsigned long events = 0;
+    label_id_t label;
+    double *time = model->params.single_sweep.trajectory.time;
+    double rec_rates[] = {0.0, 0.0};
+    double sweep_pop_sizes[] = {0.0, 0.0};
+    double event_prob, event_rand, tmp_rand, e_sum;
+    double p_coal_b, p_coal_B, total_rate, sweep_pop_tot_rate;
+    double p_rec_b, p_rec_B;
 
-    assert(num_steps > 0);
-    assert(self->num_labels == 2); /* just assuming 2 labels for now */
     ret = msp_single_sweep_initialise(self, allele_frequency[0]);
     if (ret != 0) {
         goto out;
     }
-    /* Obviously this is all utter rubbish, but hopefully these are the building
-     * blocks that you need??? */
-    for (j = 1; j < num_steps; j++) {
-        /* This will need to be some tiny value, or else the trees won't make sense */
-        self->time += 1e-6;
-        printf("j = %d ======================\n",(int)  j);
-        if (j % 3 == 1) {
-            /* Resolves to msp_std_common_ancestor_event(self, pop_id, label) */
-            ret = self->common_ancestor_event(self, 0, 1);
-        } else if (j % 3 == 2) {
-            ret = msp_single_sweep_recombination_event(self, 0, sweep_locus,
-                    allele_frequency[j]);
+
+    while (msp_get_num_ancestors(self) > 0 && self->time < max_time
+            && events < max_events && curr_step < num_steps) {
+        events++;
+        /* quick sanity check; also set pop sizes & rec_rates */
+        for (j = 0; j < self->num_labels; j++) {
+            label = (label_id_t) j;
+            num_links = fenwick_get_total(&self->links[label]);
+            ret = msp_sanity_check(self, num_links);
+            sweep_pop_sizes[j] = (double) avl_count(&self->populations[0].ancestors[label]);
+            rec_rates[j] = (double) num_links * self->recombination_rate * sweep_dt;
+            if (ret != 0) {
+                goto out;
+            }
+        }
+        curr_step++;
+        event_prob = 1.0;
+        event_rand = gsl_rng_uniform(self->rng);
+        while (event_prob > event_rand && curr_step < num_steps) {
+            sweep_dt = time[curr_step] - time[curr_step - 1];
+            /* using pop sizes grabbed from get_population_size */
+            p_coal_B = ((sweep_pop_sizes[1] * (sweep_pop_sizes[1] - 1) ) * 0.5)
+                / allele_frequency[curr_step]
+                * sweep_dt / get_population_size(&self->populations[0], time[curr_step]);
+            p_coal_b = ((sweep_pop_sizes[0] * (sweep_pop_sizes[0] - 1) ) * 0.5)
+                / (1.0 - allele_frequency[curr_step])
+                * sweep_dt / get_population_size(&self->populations[0], time[curr_step]);
+            p_rec_b = rec_rates[0] * sweep_dt;
```

Comment:
in addition I added `p_rec_b` and `p_rec_B` to make the calculation more obvious at the expense of writing a bit more code